memcache/pool_test.go

```diff
@@ -0,0 +1,239 @@
+
+		// new connection should be created
+		c2, err := p.Get(tctx(t, 100*time.Millisecond))
+		require.NoError(t, err)
+		require.NotEqual(t, c1, c2)
+		require.Equal(t, int64(2), created.Load())
+
+		// return with resumable error
+		c2.Release(ErrCacheMiss)
+
+		// new connection should not be created
+		c3, err := p.Get(tctx(t, 100*time.Millisecond))
+		require.NoError(t, err)
+		require.Equal(t, c2, c3)
+		require.Equal(t, int64(2), created.Load())
+
+		// cleanup
+		c3.Release(errors.New("dont reuse"))
+	})
+
+	t.Run("load", func(t *testing.T) {
+		created := atomic.Int64{}
+		d := net.Dialer{}
+		dctx := func(ctx context.Context, network, addr string) (net.Conn, error) {
+			created.Add(1)
+			return d.DialContext(ctx, "tcp", addr)
+		}
+
+		client := New([]string{addr.String()})
+		client.Pool = NewClusterPool(dctx, PoolConfig{MaxConns: 10})
+
+		wg := sync.WaitGroup{}
+		for i := 0; i < 100; i++ {
+			wg.Add(1)
+			go func() {
+				for j := 0; j < 1000; j++ {
+					require.NoError(t, client.Ping())
+				}
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+		require.Equal(t, int64(10), created.Load())
+	})
+
+	t.Run("stats", func(t *testing.T) {
+		p, _ := tpool(addr, PoolConfig{MaxConns: 1})
+
+		c1, err := p.Get(tctx(t, 100*time.Millisecond))
+		require.NoError(t, err)
+
+		require.Equal(t, int64(1), p.Stats().Miss)
+		require.Equal(t, int64(0), p.Stats().Hit)
+		require.Equal(t, int64(1), p.Stats().Active)
+		require.Equal(t, int64(0), p.Stats().Idle)
+
+		c1.Release(nil)
+
+		require.Equal(t, int64(0), p.Stats().Miss)
+		require.Equal(t, int64(0), p.Stats().Hit)
+		require.Equal(t, int64(0), p.Stats().Active)
+		require.Equal(t, int64(1), p.Stats().Idle)
+
+		c2, err := p.Get(tctx(t, 100*time.Millisecond))
+		require.NoError(t, err)
+
+		require.Equal(t, int64(1), p.Stats().Hit)
+		require.Equal(t, int64(0), p.Stats().Miss)
+		require.Equal(t, int64(1), p.Stats().Active)
+		require.Equal(t, int64(0), p.Stats().Idle)
+
+		// cleanup
+		c2.Release(errors.New("dont reuse"))
+	})
+}
+
+func tpool(addr net.Addr, config PoolConfig) (*Pool, *atomic.Int64) {
+	created := atomic.Int64{}
+	d := net.Dialer{}
+	dctx := func(ctx context.Context, network, addr string) (net.Conn, error) {
+		created.Add(1)
+		return d.DialContext(ctx, "tcp", addr)
+	}
+
+	pool := NewPool(addr, dctx, config)
+	return pool, &created
+}
+
+func tctx(t *testing.T, d time.Duration) context.Context {
+	ctx, cancel := context.WithTimeout(context.Background(), d)
+	t.Cleanup(cancel)
+	return ctx
+}
+
+func setupMemcached(t *testing.T) net.Addr {
+	pool, err := dockertest.NewPool("")
+	require.NoError(t, err)
+	require.NoError(t, pool.Client.Ping())
+
+	fmt.Println("Starting memcached container...")
+	resource, err := pool.RunWithOptions(&dockertest.RunOptions{
+		Repository:   "memcached",
+		Tag:          "1.6.12",
+		ExposedPorts: []string{"11211"},
+	})
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		require.NoError(t, pool.Purge(resource))
```

Comment:
There is a problem here, if you log on a finished test, `testing` won't like it. And `require` will do exactly that. In this case I'd just panic.

Reply:
fixed!